daemon/src/peer/run_state/running_waiting_room.rs

```diff
@@ -0,0 +1,345 @@
+// Linking Exception. For full terms see the included LICENSE file.
+
+use either::Either;
+use librad::{git::Urn, identities::Revision, peer::PeerId};
+use std::{
+    collections::HashMap,
+    time::{Duration, SystemTime},
+};
+
+use crate::request::SomeRequest;
+
+use super::{
+    command,
+    control,
+    waiting_room::Error as WaitingRoomError,
+    Command,
+    Event as RunStateEvent,
+    WaitingRoom,
+};
+use tokio::sync::oneshot::Sender;
+
+use serde::Serialize;
+
+/// Events that can affect the state of the waiting room
+#[derive(Debug, Clone, PartialEq, Serialize)]
+#[serde(rename_all = "camelCase", tag = "type")]
+pub enum Event {
+    /// A request was created for a urn
+    Created {
+        /// The urn bein requested
+        urn: Urn,
+    },
+    /// A query was initiated for a urn
+    Queried {
+        /// The urn bein queried
+        urn: Urn,
+    },
+    /// A peer was found who claims to have a urn
+    Found {
+        /// The urn that was found
+        urn: Urn,
+        /// The peer who claims to have it
+        peer: PeerId,
+    },
+    /// Cloning was initiated for a urn and peer
+    Cloning {
+        /// The urn we are cloning
+        urn: Urn,
+        /// The peer we are cloning from
+        peer: PeerId,
+    },
+    /// Cloning failed for a urn and peer
+    CloningFailed {
+        /// The urn that failed
+        urn: Urn,
+        /// The peer we failed to clone from
+        peer: PeerId,
+        /// A description of why the cloning failed
+        reason: String,
+    },
+    /// Cloning succeeded for a urn and peer
+    Cloned {
+        /// The urn we cloned
+        urn: Urn,
+        /// The peer we cloned from
+        peer: PeerId,
+    },
+    /// A request for a urn was canceled
+    Canceled {
+        /// The urn that was canceled
+        urn: Urn,
+    },
+    /// A request was timed out
+    TimedOut {
+        /// The urn that timed out
+        urn: Urn,
+        /// The attempts that were made before the timeout
+        attempts: Option<usize>,
+    },
+    /// One tick of the waiting room
+    Tick,
+}
+
+/// `RunningWaitingRoom` is an adapter from the interface of `WaitingRoom` to
+/// the language of commands which is spoken by `RunState`. Whenever `RunState`
+/// needs to talk to `WaitingRoom` it does so via a wrapper method on
+/// `RunningWaitingRoom`. These wrapper methods contain the logic to convert
+/// the values  returned by `WaitingRoom` methods into `Vec<Command>`.
+pub(super) struct RunningWaitingRoom {
+    waiting_room: WaitingRoom<SystemTime, Duration>,
+}
+
+impl RunningWaitingRoom {
+    pub const fn new(waiting_room: WaitingRoom<SystemTime, Duration>) -> Self {
+        Self { waiting_room }
+    }
+
+    pub fn cancel(
+        &mut self,
+        urn: Urn,
+        timestamp: SystemTime,
+        sender: Sender<Result<Option<SomeRequest<SystemTime>>, WaitingRoomError>>,
+    ) -> Vec<Command> {
```

Comment:
Not to do this with this particular PR, but:

Something that bothers me is the amalgamation of _all_ `Command`s into one structure. Something we could consider for another patch is that there is `Command` local to the `RunningWaitingRoom` and this can be added as a variant to a larger `Command` enum closer to the run state.

A `RunningWaitingRoom`s transitions and commands should be testable separated from the `RunState`.

Reply:
It's something to consider. Splitting `Command` isn't without costs. The problem is that most of the "subprotocols" of the `RunState` need to emit commands related to other subprotocols. I'm not convinced that adding indirection to make each subprotocol unaware of the other would make the code easier to follow. Regardless, probably the best way to do this would be to first split each subprotocol into it's own bit of state as I've done here, and then split the `Command` to see how it looks.

As to the testing point, it feels to me like the natural place to test this stuff is on the `RunState`, that way we're free to change implementations significantly (as this PR does) without having to update all our tests. I have a bias towards more integration-ey tests though.